File: elf/archive-file.cc
#include "mold.h"

namespace mold::elf {

struct ArHdr {
  char ar_name[16];
  char ar_date[12];
  char ar_uid[6];
  char ar_gid[6];
  char ar_mode[8];
  char ar_size[10];
  char ar_fmag[2];
};

template <typename E>
std::vector<MappedFile<Context<E>> *>
read_thin_archive_members(Context<E> &ctx, MappedFile<Context<E>> *mb) {
  u8 *begin = mb->data;
  u8 *data = begin + 8;
  std::vector<MappedFile<Context<E>> *> vec;
  std::string_view strtab;

  while (data < begin + mb->size) {
    // Each header is aligned to a 2 byte boundary.
    if ((begin - data) % 2)
      data++;

    ArHdr &hdr = *(ArHdr *)data;
    u8 *body = data + sizeof(hdr);
    u64 size = atol(hdr.ar_size);

    // Read a string table.
    if (!memcmp(hdr.ar_name, "// ", 3)) {
      strtab = {(char *)body, size};
      data = body + size;
      continue;
    }

    // Skip a symbol table.
    if (!memcmp(hdr.ar_name, "/ ", 2)) {
      data = body + size;
      continue;
    }

    if (hdr.ar_name[0] != '/')
      Fatal(ctx) << mb->name << ": filename is not stored as a long filename";

    const char *start = strtab.data() + atoi(hdr.ar_name + 1);
    std::string name(start, (const char *)strstr(start, "/\n"));
    std::string path = name.starts_with('/') ?
      name : std::string(path_dirname(mb->name)) + "/" + name;
    vec.push_back(MappedFile<Context<E>>::must_open(ctx, path));
    data = body;
  }
  return vec;
}

template <typename E>
std::vector<MappedFile<Context<E>> *>
read_fat_archive_members(Context<E> &ctx, MappedFile<Context<E>> *mb) {
  u8 *begin = mb->data;
  u8 *data = begin + 8;
  std::vector<MappedFile<Context<E>> *> vec;
  std::string_view strtab;

  while (begin + mb->size - data >= 2) {
    if ((begin - data) % 2)
      data++;

    ArHdr &hdr = *(ArHdr *)data;
    u8 *body = data + sizeof(hdr);
    u64 size = atol(hdr.ar_size);
    data = body + size;

    if (!memcmp(hdr.ar_name, "// ", 3)) {
      strtab = {(char *)body, size};
      continue;
    }

    if (!memcmp(hdr.ar_name, "/ ", 2) ||
        !memcmp(hdr.ar_name, "__.SYMDEF/", 10))
      continue;

    std::string name;

    if (hdr.ar_name[0] == '/') {
      const char *start = strtab.data() + atoi(hdr.ar_name + 1);
      name = {start, (const char *)strstr(start, "/\n")};
    } else {
      name = {hdr.ar_name, strchr(hdr.ar_name, '/')};
    }

    vec.push_back(mb->slice(ctx, name, body - begin, size));
  }
  return vec;
}

template <typename E>
std::vector<MappedFile<Context<E>> *>
read_archive_members(Context<E> &ctx, MappedFile<Context<E>> *mb) {
  switch (get_file_type(ctx, mb)) {
  case FileType::AR:
    return read_fat_archive_members(ctx, mb);
  case FileType::THIN_AR:
    return read_thin_archive_members(ctx, mb);
  default:
    unreachable(ctx);
  }
}

#define INSTANTIATE(E)                                                  \
  template std::vector<MappedFile<Context<E>> *>                        \
  read_fat_archive_members(Context<E> &, MappedFile<Context<E>> *);     \
  template std::vector<MappedFile<Context<E>> *>                        \
  read_thin_archive_members(Context<E> &, MappedFile<Context<E>> *);    \
  template std::vector<MappedFile<Context<E>> *>                        \
  read_archive_members(Context<E> &, MappedFile<Context<E>> *);

INSTANTIATE(X86_64);
INSTANTIATE(I386);
INSTANTIATE(AARCH64);

} // namespace mold::elf

File: src/archive_file.rs
//! This file contains functions to read an archive file (.a file).
//! An archive file is just a bundle of object files. It's similar to
//! tar or zip, but the contents are not compressed.
//!
//! An archive file is either "regular" or "thin". A regular archive
//! contains object files directly, while a thin archive contains only
//! pathnames. In the latter case, actual file contents have to be read
//! from given pathnames. A regular archive is sometimes called "fat"
//! archive as opposed to "thin".
//!
//! If an archive file is given to the linker, the linker pulls out
//! object files that are needed to resolve undefined symbols. So,
//! bunding object files as an archive and giving that archive to the
//! linker has a different meaning than directly giving the same set of
//! object files to the linker. The former links only needed object
//! files, while the latter links all the given object files.
//!
//! Therefore, if you link libc.a for example, not all the libc
//! functions are linked to your binary. Instead, only object files
//! that provides functions and variables used in your program get
//! linked. To make this efficient, static library functions are
//! usually separated to each object file in an archive file. You can
//! see the contents of libc.a by running `ar t
//! /usr/lib/x86_64-linux-gnu/libc.a`.

use std::path::{Path, PathBuf};

use crate::fatal;
use crate::mapped_file::MappedFile;
use crate::util;

const HEADER_SIZE: usize = 60;

/// A parsed archive member header.
struct ArHeader<'a> {
    name: &'a [u8],
    size: usize,
}

impl<'a> ArHeader<'a> {
    fn parse(bytes: &'a [u8]) -> Option<Self> {
        let bytes = bytes.get(..HEADER_SIZE)?;
        let size = parse_decimal(&bytes[48..58]);
        Some(ArHeader {
            name: &bytes[..16],
            size,
        })
    }

    fn is_strtab(&self) -> bool {
        self.name.starts_with(b"// ")
    }

    fn is_symtab(&self) -> bool {
        self.name.starts_with(b"/ ") || self.name.starts_with(b"/SYM64/ ")
    }

    /// Returns the member's file name. A BSD-style long name is stored
    /// right after the header, so `body` is advanced past it.
    fn read_name(&self, strtab: &[u8], body: &mut &'a [u8]) -> PathBuf {
        // BSD-style long filename
        if let Some(rest) = self.name.strip_prefix(b"#1/") {
            let len = parse_decimal(rest);
            let (name, remaining) = body.split_at(len.min(body.len()));
            *body = remaining;
            let name = name.split(|&b| b == 0).next().unwrap_or(&[]);
            return PathBuf::from(util::os_str(name));
        }

        // SysV-style long filename
        if let Some(rest) = self.name.strip_prefix(b"/") {
            let offset = parse_decimal(rest);
            let start = strtab.get(offset..).unwrap_or(&[]);
            let end = memchr::memmem::find(start, b"/\n").unwrap_or(start.len());
            return PathBuf::from(util::os_str(&start[..end]));
        }

        // Short fileanme
        let end = self
            .name
            .iter()
            .position(|&b| b == b'/')
            .unwrap_or(self.name.len());
        PathBuf::from(util::os_str(&self.name[..end]))
    }
}

/// Parses the leading decimal digits of a field, like `atoi`.
fn parse_decimal(bytes: &[u8]) -> usize {
    bytes
        .iter()
        .skip_while(|b| b.is_ascii_whitespace())
        .take_while(|b| b.is_ascii_digit())
        .fold(0, |acc, &b| acc * 10 + (b - b'0') as usize)
}

/// Iterates over the members of an archive as (name, body) pairs, skipping
/// the symbol table and string table.
fn archive_members(
    mf: &'static MappedFile,
    thin: bool,
) -> impl Iterator<Item = (PathBuf, &'static [u8])> {
    let data = mf.data();
    let mut pos = 8;
    let mut strtab: &'static [u8] = &[];

    std::iter::from_fn(move || {
        loop {
            if data.len() - pos < 2 {
                return None;
            }

            // Each header is aligned to a 2 byte boundary.
            if pos % 2 != 0 {
                pos += 1;
            }

            let hdr = ArHeader::parse(&data[pos..])?;
            let body_start = pos + HEADER_SIZE;
            let body_end = (body_start + hdr.size).min(data.len());
            let mut body = &data[body_start..body_end];

            // Read a string table.
            if hdr.is_strtab() {
                // Read if string table
                strtab = body;
                pos = body_end;
                continue;
            }
            // Skip a symbol table.
            if hdr.is_symtab() {
                // Skip if symbol table
                pos = body_end;
                continue;
            }

            if thin && !hdr.name.starts_with(b"#1/") && !hdr.name.starts_with(b"/") {
                fatal!(
                    "{}: filename is not stored as a long filename",
                    mf.name.display()
                );
            }

            // Read the name field
            let name = hdr.read_name(strtab, &mut body);

            if thin {
                // A thin archive member's contents live elsewhere; only a
                // BSD-style long name occupies space after the header.
                pos = body_start + (body_end - body_start - body.len());
            } else {
                pos = body_end;
            }

            // Skip BSD archive symbol tables.
            if name == Path::new("__.SYMDEF") || name == Path::new("__.SYMDEF SORTED") {
                pos = body_end;
                continue;
            }

            return Some((name, body));
        }
    })
}

/// Returns the paths of the members of a thin archive, which are stored
/// outside of the archive file, without opening them.
pub fn get_thin_archive_member_paths(mf: &'static MappedFile) -> impl Iterator<Item = PathBuf> {
    archive_members(mf, true).map(move |(name, _)| {
        if name.is_absolute() {
            name
        } else {
            mf.name.parent().unwrap_or(Path::new(".")).join(name)
        }
    })
}

pub fn read_thin_archive_members<'a>(
    chroot: &'a Path,
    mf: &'static MappedFile,
) -> impl Iterator<Item = &'static MappedFile> + 'a {
    get_thin_archive_member_paths(mf).map(move |path| mf.open_thin_member(chroot, &path))
}

pub fn read_fat_archive_members(
    mf: &'static MappedFile,
) -> impl Iterator<Item = &'static MappedFile> {
    let base = mf.data().as_ptr() as usize;
    archive_members(mf, false).map(move |(name, body)| {
        let start = body.as_ptr() as usize - base;
        mf.slice(name, start, body.len())
    })
}

pub fn read_archive_members(chroot: &Path, mf: &'static MappedFile) -> Vec<&'static MappedFile> {
    if mf.data().starts_with(b"!<arch>\n") {
        read_fat_archive_members(mf).collect()
    } else {
        debug_assert!(mf.data().starts_with(b"!<thin>\n"));
        read_thin_archive_members(chroot, mf).collect()
    }
}
